rust/blockstore/src/arrow/provider.rs

```diff
@@ -233,66 +236,39 @@ impl BlockManager {
-                                        }
-                                    }
-                                }
-                                Some(buf)
-                            }
-                            ).await;
-                            let buf =  match buf {
-                                Some(buf) => {
-                                    buf
-                                }
-                                None => {
-                                    return None;
-                                }
-                            };
-                            tracing::info!("Read {:?} bytes from s3", buf.len());
-                            let deserialization_span = tracing::trace_span!(parent: Span::current(), "BlockManager deserialize block");
-                            let block = deserialization_span.in_scope(|| Block::from_bytes(&buf, *id));
-                            match block {
-                                Ok(block) => {
-                                    self.block_cache.insert(*id, block.clone());
-                                    Some(block)
-                                }
-                                Err(e) => {
-                                    // TODO: Return an error to callsite instead of None.
-                                    tracing::error!(
-                                        "Error converting bytes to Block {:?}/{:?}",
-                                        key,
-                                        e
-                                    );
-                                    None
-                                }
-                            }
-                        },
+                let key = format!("block/{}", id);
+                // Clone the cache is cheap since it is arc type.
+                let cache_clone = self.block_cache.clone();
+                let id_copy = id.clone();
+                let cb = move |buf: Vec<u8>| async move {
+                    let deserialization_span = tracing::trace_span!(parent: Span::current(), "BlockManager deserialize block");
+                    let block = deserialization_span.in_scope(|| Block::from_bytes(&buf, id_copy));
+                    match block {
+                        Ok(block) => {
+                            cache_clone.insert(id_copy, block.clone());
+                        }
                         Err(e) => {
-                            tracing::error!("Error reading block from storage: {}", e);
-                            None
+                            tracing::error!("Error converting bytes to Block {:?}", e);
+                            return Err(Box::new(
+                                NetworkAdmissionControlError::DeserializationError,
+                            ));
                         }
                     }
+                    Ok(())
+                };
+                match self.network_admission_control.get(key, cb).await {
```

Comment:
From a design standpoint is it odd that the code here is "network admission controller?" I wonder if it really should be storage and NAC is transparent

Reply:
After thinking on it, I think my proposal would be to treat this more as a Storage Wrapper, which I mention in the bottom of the stack PR review.